Add better hint in the doctor for Maven workspaces

tgodzik opened the pull request:
Previously we recommended rerunning another import for maven workspaces, which would not help. Now we include a link to the full documentation

We will need to add a full documentation during the release - will take care of it.

![image](https://user-images.githubusercontent.com/3807253/59040847-d060c780-8877-11e9-99bf-1536059c04be.png)

olafurpg commented on metals/src/main/scala/scala/meta/internal/metals/Doctor.scala:
```diff
@@ -108,12 +108,19 @@ final class Doctor(
+    def hint() =
+      if (isMaven)
+        "check how to enable semanticDB on the " +
+          "<a href=https://scalameta.org/metals/docs/build-tools/overview.html>metals website</a>"
```
```suggestion
          "<a href=https://scalameta.org/docs/semanticdb/guide.html#producing-semanticdb>Scalameta website</a>"
```

tgodzik replied:
I was thinking of making Maven specific instructions for this. There is nothing on the scalameta website about maven.

olafurpg replied:
Should we create a Maven-specific page in that case? Similar to the sbt page https://scalameta.org/metals/docs/build-tools/sbt.html

tgodzik replied:
That's my thinking yes.

olafurpg commented on metals/src/main/scala/scala/meta/internal/metals/Doctor.scala:
```diff
@@ -108,12 +108,19 @@ final class Doctor(
+    def isMaven: Boolean = workspace.resolve("pom.xml").isFile
+    def hint() =
+      if (isMaven)
+        "check how to enable semanticDB on the " +
```
```suggestion
        "Enable SemanticDB following instructions on the " +
```

tgodzik replied:
fixed, but `enable` not `Enable` since I capitalize below if this is the only sentence.

tgodzik commented:
@olafurpg I added a new section about maven, let me know if there is anything to change.

Should we add similar sections for `gradle`  and `mill` ?

olafurpg commented on docs/build-tools/maven.md:
```diff
@@ -0,0 +1,80 @@
+    <dependency>
+      <groupId>org.scalameta</groupId>
+      <artifactId>semanticdb-scalac_${scala.version}</artifactId>
+      <version>4.1.4</version>
```
```suggestion
      <version>@@SCALAMETA_VERSION@@</version>
```

olafurpg commented on docs/build-tools/maven.md:
```diff
@@ -0,0 +1,80 @@
+    <version>4.0.2</version>
+    <configuration>
+      <args>
+        <arg>-Xplugin:@@replace_with_semanticDB_path@@</arg>
```
Using the absolute local path seems fragile, is there a Scala maven plugin that makes it easier to add a compiler plugin?

olafurpg replied:
scalor-maven-plugin: https://random-maven.github.io/scalor-maven-plugin/2.12/eclipse-config-mojo.html#definePluginList

scala-maven-plugin: http://davidb.github.io/scala-maven-plugin/example_scalac_plugins.html (see "using a scalac plugin"

tgodzik replied:
Och alright, I haven't noticed this solution. That makes it much easier now.

olafurpg commented on docs/build-tools/maven.md:
```diff
@@ -0,0 +1,80 @@
+        <arg>-Xplugin:@@replace_with_semanticDB_path@@</arg>
+        <arg>-P:semanticdb:synthetics:on</arg>
+        <arg>-P:semanticdb:failures:warning</arg>
+        <arg>-P:semanticdb:sourceroot:${project.basedir}</arg>
```
Is `project.basedir` the root directory of the build or the base directory of the project? I'm wondering if there is actually such as thing as `baseDirectory.in(ThisBuild).value` in Maven 🤔 We might need built-in support for automatically setting the sourceroot in Bloop, cc/ @jvican

tgodzik replied:
`project.basedir: The directory that the current project resides in.`

olafurpg replied:
For most common Maven builds I suspect we need the sourceroot to match the parent of `project.basedir`.

tgodzik replied:
Plugin is usually defined in the root project, so this should point to the correct location.

olafurpg replied:
Have you double checked? It might be worth adding a test case to confirm

tgodzik replied:
Changed to `maven.multiModuleProjectDirectory` - that seems to be the root dir for all of the projects. Checked it and looks to be working.

tgodzik commented:
> Adding a new page for Maven is a great idea! We should probably do the same for Gradle and Mill.
> 
> We should also update the build tools tables as was done in #742. The "installation" column for Maven should be something like  instead of  with an explanation below " Requires manual changes to the build to enable 'Find symbol references'"



> Adding a new page for Maven is a great idea! We should probably do the same for Gradle and Mill.
> 
> We should also update the build tools tables as was done in #742. The "installation" column for Maven should be something like  instead of  with an explanation below " Requires manual changes to the build to enable 'Find symbol references'"

I added another column and switched the green check marks to be fully working features, while white one to be partly working. Green seems to me to show more positive sentiment. What do you think?

Also fixed the maven part, it's much simple rnow, thanks! I will add gradle and mill pages tomorrow.

olafurpg commented on docs/build-tools/maven.md:
```diff
@@ -0,0 +1,61 @@
+title: Maven
+---
+
+Maven is one of the most common build tools in the JVM ecosystem and it also allows for using scala through the [scala-maven-plugin](https://davidb.github.io/scala-maven-plugin/usage.html).
```
```suggestion
Maven is one of the most common build tools in the JVM ecosystem and it also allows for using Scala through [scala-maven-plugin](https://davidb.github.io/scala-maven-plugin/usage.html) and [scalor-maven-plugin](https://github.com/random-maven/scalor-maven-plugin).
```

tgodzik replied:
I don't think bloop plugin works with scalor, so maybe we shouldn't mention it?

olafurpg replied:
I didn't know. Is there a something we can link to for scalor-maven-plugin users? It seems like an actively maintained project with fairly rich features. If so, we can add a new sentence
```
also allows for using scala through the [scala-maven-plugin](https://davidb.github.io/scala-maven-plugin/usage.html).
The [scalor-maven-plugin](https://github.com/random-maven/scalor-maven-plugin) is not supported, see $LINK for more information.
```

tgodzik replied:
> I didn't know. Is there a something we can link to for scalor-maven-plugin users?

I don't think we have anything to point to right now - maybe add a comment that we would appreciate a contribution to bloop?

olafurpg replied:
Fair enough, we can skip mentioning scalor-maven-plugin in that case and if somebody asks we'll point to Bloop.

olafurpg commented on docs/build-tools/maven.md:
```diff
@@ -0,0 +1,61 @@
+The first time you open Metals in a new workspace it prompts you to import the
+build. Select "Import build" to start automatic installation.
+
+This will create all the needed bloop files, however there will be a warning, since the semanticDB plugin was not added yet in the automatic import. This will be added later. Most features should work without it, however some require semanticDB files to be provided alongside compiled data. To do that we need a couple of steps that are explained in the manual instalation section.
```
I recommend formatting markdown files with prettier even if we don't enforce this in CI. Prettier provides automatic line-wrapping which makes reviewing a bit easier.

olafurpg replied:
```suggestion
This will create all the needed Bloop files, however there will be a warning, since the SemanticDB plugin was not added yet in the automatic import. This will be added later. Most features should work without it, however some require SemanticDB files to be provided alongside compiled data. To do that we need a couple of steps that are explained in the manual installation section.
```

olafurpg commented on docs/build-tools/maven.md:
```diff
@@ -0,0 +1,61 @@
+
+## Manual instalation
+
+First, we need to add a couple of options to the scala compiler in the configuration section:
```
```suggestion
First, we need to add a couple of options to the Scala compiler in the configuration section:
```

olafurpg commented on docs/build-tools/maven.md:
````diff
@@ -0,0 +1,61 @@
+</plugin>
+```
+
+`@@SCALAMETA_VERSION@@` version can be replaced with whatever is the newest semanticDB version.
````
```suggestion

```

Not needed, `@@SCALAMETA_VERSION@@` is automatically replaced by mdoc.

olafurpg commented on docs/build-tools/maven.md:
```diff
@@ -0,0 +1,61 @@
+
+Next, we need to do run bloopInstall via maven, which can be done easily through:
+
+`mvn ch.epfl.scala:maven-bloop_2.10:1.3.0:bloopInstall -DdownloadSources=true`
```
```suggestion
`mvn ch.epfl.scala:maven-bloop_2.10:@@BLOOP_VERSION@@:bloopInstall -DdownloadSources=true`
```

olafurpg commented on docs/build-tools/maven.md:
```diff
@@ -0,0 +1,61 @@
+
+`mvn ch.epfl.scala:maven-bloop_2.10:1.3.0:bloopInstall -DdownloadSources=true`
+
+where `1.3.0` can be replaced with newer bloop version.
```
```suggestion
```

olafurpg commented on docs/build-tools/maven.md:
```diff
@@ -0,0 +1,61 @@
+
+If you don't want to modify the `pom.xml` you can also run bloopInstall with an additional parameter:
+
+`mvn ch.epfl.scala:maven-bloop_2.10:1.3.0-RC1:bloopInstall -DdownloadSources=true -DaddScalacArgs=-Xplugin:@@semantic_DB_plugin_jar@@|-P:semanticdb:synthetics:on|-P:semanticdb:failures:warning|-P:semanticdb:sourceroot:/home/tgodzik/Documents/maven-test-repo|-Yrangepos|-Xplugin-require:semanticdb'`
```
````suggestion
```
mvn ch.epfl.scala:maven-bloop_2.10:@@BLOOP_VERSION@@:bloopInstall -DdownloadSources=true -DaddScalacArgs=-Xplugin:/path/to/semanticdb-scalac.jar|-P:semanticdb:synthetics:on|-P:semanticdb:failures:warning|-P:semanticdb:sourceroot:/path/to/workspace|-Yrangepos|-Xplugin-require:semanticdb'
```
````

olafurpg commented on docs/build-tools/maven.md:
```diff
@@ -0,0 +1,61 @@
+
+`-DaddScalacArgs` takes a string with additional scalac options separated by `|`. You would need to download the correct plugin first and manually replace `@@semantic_DB_plugin_jar@@` with the path to the semanticDB plugin jar.
+
+If you choose this option though you should disable automatic import, because it will override the changes.
```
```suggestion
If you choose this option though you should select "Don't show again" when Metals prompts to import the build.
```

olafurpg commented on docs/build-tools/overview.md:
```diff
@@ -31,14 +32,20 @@ compiler plugin and `-Yrangepos` option enabled.
 
 ## Goto library dependencies
 
-**✅**: it is possible to navigate Scala+Java library dependencies using "Goto
```
What do you think about replacing the checkboxes with words?

✅ becomes "Semi-automatic"

✔️ becomes "Automatic"

tgodzik replied:
We can do that, I was just thinking the idea was to use those ticks?

olafurpg replied:
Seeing the table now, I'm not convinced which one of ✔️ and ✅ intuitively means more complete than the other. I think the words "Automatic" and "Semi-automatic" are more self-explanatory.

tgodzik replied:
Changed - let me know what you think

olafurpg commented on docs/build-tools/mill.md:
```diff
@@ -0,0 +1,34 @@
+
+`mill mill.contrib.Bloop/install`
+
+Afterwards just can just open metals and start working on your code.
```
```suggestion
Afterwards just can just open Metals and start working on your code.
```

olafurpg commented on docs/build-tools/mill.md:
```diff
@@ -0,0 +1,34 @@
+Remember to replace the `VERSION` with your mill version.
+
+After adding the line you should be able to generate Bloop config files needed
+to work with metals using the below command:
```
```suggestion
to work with Metals using the below command:
```
I may be have missed some others. We should also try to use "Mill" consistently instead of "mill"

olafurpg commented on docs/build-tools/gradle.md:
```diff
@@ -0,0 +1,96 @@
+            scalaCompilerPlugin
+        }
+        dependencies {
+            scalaCompilerPlugin "org.scalameta:semanticdb-scalac_$scalaVersion:$scalaMetaVersion"
```
```suggestion
            scalaCompilerPlugin "org.scalameta:semanticdb-scalac_$scalaVersion:@@SCALAMETA_VERSION@@"
```

olafurpg commented on docs/build-tools/gradle.md:
```diff
@@ -0,0 +1,96 @@
+        mavenCentral()
+    }
+    dependencies {
+        classpath 'ch.epfl.scala:gradle-bloop_2.11:${BuildInfo.gradleBloopVersion}'
```
```suggestion
        classpath 'ch.epfl.scala:gradle-bloop_2.11:@@BLOOP_VERSION@@'
```

olafurpg commented on docs/build-tools/gradle.md:
```diff
@@ -0,0 +1,96 @@
+
+Now we can run `gradle bloopInstall`, which will create a Bloop configuration
+files. This will enable us to work with metals and most features will work, but
+for everything to work properly we need to also add the semanticDB plugin. This
```
```suggestion
for everything to work properly we need to also add the SemanticDB plugin. This
```

olafurpg commented on docs/build-tools/overview.md:
```diff
@@ -31,14 +32,20 @@ compiler plugin and `-Yrangepos` option enabled.
+
+**Automatic**: it is possible to find all references to a symbol in the project
+
+**Semi-automatic**: it is possible to find references as soon as we configure additional SemanticDB plugin for the compiler, check separate build tool pages for details
```
```suggestion
**Semi-automatic**: it is possible to 'Find symbol references' as soon the SemanticDB compiler plugin is manually enabled in the build, check separate build tool pages for details.
```
